File: py/check.py
import mysql.connector
from mysql.connector import Error
import time
from datetime import datetime

def connect():
    try:
        conn = mysql.connector.connect(
            host="localhost",
            user="root",
            password="",
            database="mraweb"
        )
        if conn.is_connected():
            return conn
    except Error:
        print("⚠️ PHPMyAdmin/MySQL belum run! Sila pastikan XAMPP atau MySQL server dijalankan.")
        return None

def check1():
    db = connect()
    if db is None:
        return []  # tiada data jika database tidak connect
    cursor = db.cursor()
    cursor.execute("SELECT id, lponum, duedate FROM projek")
    result = cursor.fetchall()
    db.close()
    return result

def updatebildate():
    db = connect()
    if db is None:
        return  # keluar jika tiada sambungan database
    cursor = db.cursor()
    data = check1()
    for row in data:
        id = row[0]
        lponum = row[1]
        duedate = row[2]
        today = datetime.now().date()
        end = duedate.date()
        total = (end - today).days
        if total <= -1:
            print(f"ID: {id}, LPO Number: {lponum}, Tarikh akhir: {end}, Tarikh hari ni: {today}, Jumlah hari: {0} days")
            cursor.execute("UPDATE projek SET bildate = %s WHERE id = %s", (0, id))
        else:
            print(f"ID: {id}, LPO Number: {lponum}, Tarikh akhir: {end}, Tarikh hari ni: {today}, Jumlah hari: {total} days")
            cursor.execute("UPDATE projek SET bildate = %s WHERE id = %s", (total, id))
    db.commit()
    db.close()

def main():
    while True:
        print(f"Check jumlah hari akhir projek.")
        updatebildate()
        time.sleep(3)

if __name__ == "__main__":
    main()

File: scala/check.scala
import java.sql.{Connection, DriverManager, SQLException}
import java.time.LocalDate
import java.time.temporal.ChronoUnit

object Check {

    val url = "jdbc:mysql://localhost:3306/mraweb"
    val username = "root"
    val password = ""

    // load driver SEKALI
    Class.forName("com.mysql.cj.jdbc.Driver")

    def getConnection(): Option[Connection] = {
        try {
            Some(DriverManager.getConnection(url, username, password))
        } catch {
            case _: SQLException =>
                println("❌ XAMPP / MySQL belum hidup")
                None
        }
    }

    def updatedata(id: String, baki: Long, conn: Connection): Unit = {
        val sql = "UPDATE projek SET bildate = ? WHERE id = ?"
        val stm = conn.prepareStatement(sql)
        stm.setLong(1, baki)
        stm.setString(2, id)
        stm.executeUpdate()
        stm.close()
    }

    def main(args: Array[String]): Unit = {

        while (true) {

            getConnection() match {
                case Some(conn) =>
                    println("✅ MySQL bersambung")

                    val stmt = conn.createStatement()
                    val rs = stmt.executeQuery("SELECT id, duedate FROM projek")

                    val hariIni = LocalDate.now()

                    while (rs.next()) {
                        val id = rs.getString("id")
                        val dateSql = rs.getDate("duedate")

                        if (dateSql != null) {
                            val tarikhAkhir = dateSql.toLocalDate
                            val bakiHari = ChronoUnit.DAYS.between(hariIni, tarikhAkhir)
                            updatedata(id, bakiHari, conn)
                        }
                    }

                    rs.close()
                    stmt.close()
                    conn.close()

                case None =>
                    println("⏳ Menunggu XAMPP dihidupkan...")
            }

            Thread.sleep(3000)
        }
    }
}